tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,201 @@
+        assert_script_run "ssh root\@$guest \"systemctl stop irqbalance.service\"" if script_run("ssh root\@$guest \"systemctl status irqbalance\"") eq 0;
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        # Bind all NIC IRQs to cpu0 in guest
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+        }
+        my @affinities_with_binding_cpu0 = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        foreach (@affinities_with_binding_cpu0) {
+            die "Failed to bind NIC IRQs to CPU0" if $_ ne 1;
+        }
+
+        # Start the irqbalance.service again
+        assert_script_run("ssh root\@$guest \"systemctl start irqbalance\"");
+
+        # Check if the SMP affinities are distributed among CPUs
+        my @affinities_with_irqbalance = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+        # So it is correct if some of them may happen to run on cpu0
+        # but it would not be correct if all IRQs were bound to cpu0.
+        if (guest_is_sle($guest, '<=15-sp3')) {
+            my $multiply = 1;
+            $multiply *= $_ foreach (@affinities_with_irqbalance);
+            die "SMP affinities were all bound to CPU0 with irqbalance.service enabled" if $multiply == 1;
+        }
+        # for SLE15SP4 (or newer release), IRQs are specified on all CPU cores
+        # fg. affinity value is '0xf' for machine with 4 CPU cores
+        else {
+            foreach (@affinities_with_irqbalance) {
+                die "It is not correct that one of the NIC IRQs was bound to cpu0 with irqbalance enabled." if $_ eq 1;
+            }
+        }
+
+        # Get the network interrupts distribution over CPUs in guest prior to network loads
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+
+        # Generate network load in guest
+        assert_script_run("ssh root\@$guest \"ping -f -c 500 download.opensuse.org\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
```

Comment:
I do not think this is accurate or even close to accurate under certain circumstances. I am afraid there is continuous traffic between SUT and worker, so it is not very meaningful to do this judgement "if $increased_irqs_on_cpu[$cpu_id] == 0". I think you'd better give it a larger number based on the volume of traffic generated and do comparison like "if $increased_irqs_on_cpu[$cpu_id] < 50" ?

Reply:
To be more accurate, I think there might be interrupt queues per device, so you can assign different applications to different queues to separate interrupts.

Reply:
sounds complex, I'd rather do some reseach on the value.

Reply:
This is irqbalance test. IMHO, it does not matter from which application or where the NIC traffic are from. What matters is that the network traffic is balanced on all cpus. Once the triggered network traffic is large enough, if irqbalance is working well, statiscally, all cpus have large enough possibility( quite close to 99.9...% ) to get at least one network packet handled. That's what we want to verify to ensure irqbalance works -- packets are being distributed.

Reply:
To be honest, "if $increased_irqs_on_cpu[$cpu_id] == 0" does not reflect the real situation, especially for cpu0, because it always gets traffic.

Reply:
I read through the test process again, and think about above comment, I think:
1) comparing value with 0 is correct, and not weak.
why? 
$increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id]

total_irqs_on_cpu_after_network_download is got after " bind irq to cpu0 -> restart irqbalance service in guest", and at this time, if the balance service is working, all interrupts are being distributed (not binded on cpu0).
It does not matter whether the interrupts comes from the ping that we intended to trigger, or the backgorund interrupts on test bed. What matters is that these interrupts are being distributed and will be counted in our later check of the diff. I guess this is where you kind of get lost or misunderstand.
Why we need to trigger large enough ping is that we do not know how much test bed background interrupts are here, and we are afraid that it is not enough to LET ALL CPU HAVE AT LEAST ONE INTERRUPT. It is possible to have (100, 200,300,0) with different balancing algorithm. So only when ping+background interrupts on test bed is large enough, statistically, we can ensure that each cpu has at least 1 interrupt.
So $increased_irqs_on_cpu[$cpu_id] == 0 check is not weak.

2) you made a good point about "fair balancing".
Why it is not included in the design is that there are different balancing algorithms, the "fair" is different in each different algorithm. Our goal of this automation is to check the irq is balancing, rather than validating that each balancing algorithm is behaving correctly(50% fair, or 90% fair). So we do not include this check point.

Reply:
I always understand the test logic. As I said, there is no test isolation at all and no deep dive into irqbalance at all either, so the checkpoint is quite weak from this point of view. Actually what is being checked here is not quite relevant to what is generated as long as there is enough volume, so it is really weak. I said it is weak without criticism. @Julie-CAO can have her own decision if she think more delicate checking is not needed.

Reply:
Well, I did not intend to reinforce my opinions on anyone. What I was doing was to explain the test design and double confirm that at least within our team, we do not see obvious flaws here(actually we also invited the dev team to review the design and they agreed). Based on your reply, it seems there is still gap. @Julie-CAO  @waynechen55  How about we three has a short call for this? We need to deliver the right thing. Discussion is a good way to achieve this.

Reply:
we agreed that this design can meet our test requirement, so let's keep this simple design.